aeppl/opt.py

```diff
@@ -178,8 +158,84 @@ def naive_bcast_rv_lift(fgraph, node):
     return [bcasted_node.outputs[1]]
 
 
-logprob_canonicalize = SequenceDB()
+class ScanRV(RandomVariable):
+    """A placeholder used to specify a log-likelihood for a scan sub-graph."""
+
+    @classmethod
+    def create_node(cls, node, scan_rvs):
+        out_var = node.default_output()
+
+        scanrv_op = ScanRV(
+            f"{out_var.name if out_var.name else ''}-scan_rv",
+            out_var.ndim,
+            [_rvs.out.ndim for _rvs in scan_rvs],
+            out_var.dtype,
+            inplace=False,
+        )
+        scanrv_op.num_indices = len(scan_rvs)
+
+        new_node = scanrv_op.make_node(None, node.out.shape, None, *scan_rvs)
+
+        return new_node
+
+    def _infer_shape(self, size, dist_params, param_shapes=None):
+        return size
+
+    def rng_fn(self, rng, *args, **kwargs):
+        raise NotImplementedError()
+
+
+@local_optimizer([Scan])
+def scan_replace(fgraph, node):
+
+    node_outs = node.op.outputs
+
+    if not node_outs:
+        return
+
+    inner_rvs = []
+
+    # GWalk the inner fgraphs for any random variables
+    for node_out in node_outs:
+        out_fgraph = FunctionGraph(
+            outputs=[node_out],
+            clone=True,
+            copy_orphans=False,
+            copy_inputs=False,
+            features=[ShapeFeature()],
+        )
+
+        for _out in out_fgraph.toposort():
+            if isinstance(_out.op, RandomVariable):
+                inner_rvs.append(_out)
+
+    # If no RV found in inner graph then return
+    if not inner_rvs:
+        return
+
+    # Replace this sub-graph with a `ScanRV`
+    new_node = ScanRV.create_node(node, inner_rvs)
+
+    new_scan_rv = new_node.default_output()
+    new_scan_rv.name = "scan"
+
+    # FIXME: This is pretty hackish
+    fgraph.import_node(new_node, import_missing=True, reason="scan_rv")
+
+    return [new_scan_rv]
```

Comment:
So as my first try I did something similar to what we do with `MixtureRV`. (And, yes, it doesn't work)

This replacement gets stuck in some long 'optimization' computation.

Reply:
The replacement performed by `scan_replace` should only occur when the inner-outputs that determine the outer-outputs are all `RandomVariable` nodes (i.e. the (relevant) outputs of scan are effectively a sequence of `RandomVariable` outputs), because that's when we can compute the log-probability with something like `at.sum(scan(lambda ..., value, ...: logprob(inner_out, value), ...))`.

The other rewrites are applied separately to the original `Scan`&mdash;through the normal course of running `logprob_rewrites_db`&mdash;until the inner-graph's `RandomVariable`s are pushed down enough to become the inner and/or outer-outputs.  `scan_replace` is only there to check that this has been done sufficiently enough to replace the `Scan`.